File: standalone/runner.pardiso.cpp
/*******************************************************************************
 * Copyright (C) 2025 Theodore Chang
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 ******************************************************************************/

#include <mpl/mpl.hpp>

int main(int, char**) {
    constexpr int NUM_NODE = 2;

    const auto& comm_world{mpl::environment::comm_world()};
    const auto worker = comm_world.spawn(0, NUM_NODE, {"solver.pardiso"});
    const auto all = mpl::communicator(worker, mpl::communicator::order_low);

    int config[7]{};
    int iparm[64]{};

    iparm[0] = 1;   // solver default parameters overriden with provided by iparm
    iparm[1] = 3;   // use METIS for fill-in reordering
    iparm[5] = 0;   // write solution into x
    iparm[7] = 2;   // max number of iterative refinement steps
    iparm[9] = 13;  // perturb the pivot elements with 1E-13
    iparm[10] = 1;  // use nonsymmetric permutation and scaling MPS
    iparm[12] = 1;  // switch on Maximum Weighted Matching algorithm (default for non-symmetric)
    // iparm[17] = -1; // output: Number of nonzeros in the factor LU
    // iparm[18] = -1; // output: Mflops for LU factorization
    iparm[26] = 0;  // check input data for correctness
    iparm[39] = 0;  // input: matrix/rhs/solution stored on master

    constexpr int N = 100, NRHS = 1;

    config[0] = 11;   // mtype
    config[1] = NRHS; // nrhs
    config[2] = 1;    // maxfct
    config[3] = 1;    // mnum
    config[4] = 0;    // msglvl
    config[5] = N;    // n
    config[6] = N;    // nnz

    std::vector<int> ia(N + 1), ja(N);
    std::vector<double> a(N), b(N * NRHS, 1.);

    for(auto i = 0; i < N; i++) ia[i] = ja[i] = a[i] = i + 1;
    ia[N] = N + 1;

    all.bcast(0, config);
    all.bcast(0, iparm);

    mpl::irequest_pool requests;

    requests.push(worker.isend(ia, 0, mpl::tag_t{0}));
    requests.push(worker.isend(ja, 0, mpl::tag_t{1}));
    requests.push(worker.isend(a, 0, mpl::tag_t{2}));
    requests.push(worker.isend(b, 0, mpl::tag_t{3}));

    requests.waitall();

    int error = -1;
    worker.recv(error, 0);
    if(0 == error) worker.recv(b, 0);

    for(auto i = 0; i < b.size(); i++) printf("x[%ld] = %+.8f\n", i, b[i]);

    return 0;
}

File: standalone/runner.pardiso.py
#  Copyright (C) 2025 Theodore Chang
#
#  This program is free software: you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation, either version 3 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program.  If not, see <https://www.gnu.org/licenses/>.

#
# This script runs a parallel solver using MPI to distribute the computation.
#
# Functions:
#     run(nprocs: int, N: int, NRHS: int):
#         Spawns the solver process, broadcasts the problem configuration, sends the matrices,
#         receives the error code and solution, and disconnects the worker and merged communicator.
#
# Usage:
#     runner.pardiso.py <nprocs> <N> <NRHS>
#
# Arguments:
#     nprocs (int): Number of processes to spawn.
#     N (int): Size of the matrix.
#     NRHS (int): Number of right-hand sides.
#
# Example:
#     python runner.pardiso.py 4 100 10
#

import numpy
import sys
from array import array
from mpi4py import MPI


def run(nprocs: int, N: int, NRHS: int):
    comm = MPI.COMM_WORLD

    # spawn the solver
    worker = comm.Spawn("./solver.pardiso", maxprocs=nprocs)
    all = worker.Merge()

    # broadcast the problem configuration
    all.Bcast(array("i", [11, 0, N, N, NRHS]), root=0)
    all.Bcast(numpy.zeros(64, dtype=numpy.int32), root=0)

    # send the matrices
    ia = numpy.array([x + 1 for x in range(N + 1)], dtype=numpy.int32)
    ja = numpy.array([x + 1 for x in range(N)], dtype=numpy.int32)
    a = numpy.array([x + 1 for x in range(N)], dtype=numpy.float64)
    b = numpy.ones((N, NRHS), dtype=numpy.float64, order="F")
    for i in range(NRHS):
        b[:, i] = i + 1

    worker.Send(ia, dest=0, tag=0)
    worker.Send(ja, dest=0, tag=1)
    worker.Send(a, dest=0, tag=2)
    worker.Send(b, dest=0, tag=3)

    # receive the error code and the solution if no error
    error = array("i", [-1])
    worker.Recv(error)
    if 0 == error[0]:
        X = numpy.ones((N, NRHS), dtype=numpy.float64, order="F")
        worker.Recv(X)
        print("Solution:\n", X)

    worker.Disconnect()
    all.Disconnect()


if __name__ == "__main__":
    if len(sys.argv) < 4:
        print("Usage: runner.pardiso.py <nprocs> <N> <NRHS>")
        sys.exit(1)

    nprocs = int(sys.argv[1])
    N = int(sys.argv[2])
    NRHS = int(sys.argv[3])

    run(nprocs, N, NRHS)
